### ISSUE TITLE
NeurIPS 2024 | SmallToLarge (S2L): Scalable Data Selection for Fine-tuning Large Language Models by Summarizing Training Trajectories of Small Models

### ISSUE THREAD

**Issue by chufanchen** [chufanchen/read-paper-and-code#186]
Code: https://github.com/BigML-CS-UCLA/S2L
Paper: https://arxiv.org/abs/2403.07384

[Training Trajectories of Language Models Across Scales(ACL 2023)](https://arxiv.org/abs/2212.09803) demonstrates that the training dynamics of most examples are consistent across differently sized models of the same family, and this phenomena even generalizes across different model families.

**_Theorem_**: If examples $i$ and $j$ have similar loss trajectories on the proxy model, i.e., $\| \mathbf{L}_i^{\text {proxy }}$ $\mathbf{L}_j^{\text {proxy }} \| \leq \epsilon$, and their loss trajectories on the proxy and target model is similar, i.e., $\| \mathbf{L}_p^{p r o x y}-$ $\mathbf{L}_p^{\text {target }} \| \leq \delta$ for $p \in\{i, j\}$, then $i$ and $j$ have similar gradients throughout training the target model:

```math
\left\|\nabla \mathcal{L}_i^{\text {target }}(\boldsymbol{\theta})-\nabla \mathcal{L}_j^{\text {target }}(\boldsymbol{\theta})\right\| \leq \frac{2 \epsilon^{\prime}+2 C D^2}{d}=\Delta .
```

where $\epsilon^{\prime}=\epsilon+2 \delta$ and $\|\boldsymbol{\theta}\| \leq D$ for all $t$.

**_Corollary_**: Under the assumptions of Theorem, applying Incremental Gradient(IG) with stepsize $\alpha$ to subsets found by _S2L_, converges to the neighborhood of the optimal solution, as follows:

```math
\left\|\boldsymbol{\theta}^{t+1}-\boldsymbol{\theta}^*\right\|^2 \leq(1-\alpha c)^{t+1}\left\|\boldsymbol{\theta}^t-\boldsymbol{\theta}^*\right\|^2+2 \xi R / c^2+\alpha B^2\left(r_{\min } / k\right)^2 \boldsymbol{g}_{\max }^2
```

where $c \leq\vert\boldsymbol{H}\vert, B=k \cdot K$ is the total size of the subset, $g_{\max}$ is the largest gradient norm of individual examples during training, $r_{\min}=\min_{j}\vert C_j\vert$, $r_{\max}=\max_{j}\vert C_j\vert$, $R=\min (d_0, B g_{\max}+\xi / c)$ and $d_0=\vert \theta^{0}-\theta^{\star} \vert$ is the initial distance to the optimal solution $\theta^*$, and $\xi$ is given by:

```math
\xi=K\left[r_{\min } \Delta+\left(r_{\max }-r_{\min }\right) \boldsymbol{g}_{\max }\right]
```
